server/src/main/java/org/elasticsearch/index/engine/InternalEngine.java

```diff
@@ -1256,12 +1236,11 @@ protected final DeletionStrategy planDeletionAsNonPrimary(Delete delete) throws
             // See testRecoveryWithOutOfOrderDelete for an example of peer recovery
             plan = DeletionStrategy.processButSkipLucene(false, delete.seqNo(), delete.version());
         } else {
-            final OpVsLuceneDocStatus opVsLucene = compareOpToLuceneDocBasedOnSeqNo(delete);
-            if (opVsLucene == OpVsLuceneDocStatus.OP_STALE_OR_EQUAL) {
-                plan = DeletionStrategy.processAsStaleOp(softDeleteEnabled, false, delete.seqNo(), delete.version());
+            final OptionalLong prevSeqNo = resolveDocSeqNo(delete);
+            if (prevSeqNo.isPresent() == false || prevSeqNo.getAsLong() < delete.seqNo()) {
+                plan = DeletionStrategy.processNormally(prevSeqNo.isPresent() == false, delete.seqNo(), delete.version());
             } else {
-                plan = DeletionStrategy.processNormally(opVsLucene == OpVsLuceneDocStatus.LUCENE_DOC_NOT_FOUND,
-                    delete.seqNo(), delete.version());
+                plan = DeletionStrategy.processAsStaleOp(softDeleteEnabled, false, delete.seqNo(), delete.version());
```

Comment:
same comment about an assertion loading the term if seq no is equal

Reply:
ping

Reply:
We have to wait for the actual rollback to add the assertion here. I added a TODO in resolveDocSeqNo.